images/airflow/2.9.2/python/mwaa/entrypoint.py

```diff
@@ -189,16 +189,36 @@ def create_queue() -> None:
             raise e
 
 
-def _requirements_has_constraints(requirements_file: str):
-    with open(requirements_file, "r") as file:
-        for line in file:
-            # Notice that this regex check will also match lines with commented out
-            # constraints flag. This is intentional as a mechanism for users who
-            # want to avoid enforcing the default Airflow constraints, yet don't want
-            # to provide a constraints file.
-            if re.search(r"-c |--constraint ", line):
-                return True
+def _requirements_has_constraints(
+    requirements_file: str,
+    subprocess_logger: logging.Logger,
+):
+    encodings_to_try = ["utf-8", "ascii", "latin1"]
+    for encoding in encodings_to_try:
+        try:
+            with open(requirements_file, "r", encoding=encoding) as file:
+                for line in file:
+                    # Notice that this regex check will also match lines with commented
+                    # out constraints flag. This is intentional as a mechanism for users
+                    # who want to avoid enforcing the default Airflow constraints, yet
+                    # don't want to provide a constraints file.
+                    if re.search(r"-c |--constraint ", line):
+                        return True
+            return False
+        except UnicodeDecodeError:
+            logger.warning(
+                f"Failed to decode {requirements_file} using {encoding} encoding."
+            )
 
+    # latin1 is not expected to fail, since it just reads the file as bytes (regardless
+    # of the content), so this block is not expected to be reach, but adding it to be
+    # on the safe side.
+    subprocess_logger.warning(
+        f"Failed to decode {requirements_file} using {encodings_to_try} encodings. "
+        "Cannot determine whether the requirements file has constraints or not; "
+        "enforcing Airflow constraints."
```

Comment:
nit: should we say "enforcing local Airflow constraints"? I think that's what we say in the case no constraint is defined

Reply:
I think yes, cause this is the action we are taking here when we return False. Unless you are suggesting we don't return True, and hence just enforce the essential constraints?

Reply:
Naah, default action could be what you already have. just add "local" in there : )